crates/code_generation/src/accessor.rs

```diff
@@ -1,383 +1,310 @@
-    path: &[PathSegment],
-    reference_type: ReferenceType,
-    instance: TokenStream,
-    cycler_instance_prefix: TokenStream,
-    cycler_instances: &[String],
+    prefix: TokenStream,
+    path: &Path,
+    reference_type: ReferenceKind,
+    cycler: &Cycler,
 ) -> TokenStream {
-    for segment in path.iter() {
-        if segment.is_variable && segment.name != "cycler_instance" {
-            unimplemented!("only $cycler_instance is implemented");
-        }
+    if let Some(segment) = path
+        .segments
+        .iter()
+        .find(|segment| segment.is_variable && segment.name != "cycler_instance")
+    {
+        let name = &segment.name;
+        unimplemented!("unexpected `${name}` only $cycler_instance can be used as variable in path",)
     }
-    let path_contains_variable = path.iter().any(|segment| segment.is_variable);
-    if path_contains_variable {
-        let mut token_stream = TokenStream::default();
-        token_stream.append(TokenTree::Ident(format_ident!("match")));
-        token_stream.extend(instance);
-        let mut token_stream_within_match = TokenStream::default();
-        for cycler_instance in cycler_instances {
-            token_stream_within_match.extend(cycler_instance_prefix.clone());
-            token_stream_within_match.append(format_ident!("{}", cycler_instance));
-            token_stream_within_match.append(TokenTree::Punct(Punct::new('=', Spacing::Joint)));
-            token_stream_within_match.append(TokenTree::Punct(Punct::new('>', Spacing::Alone)));
-            token_stream_within_match.extend(path_to_accessor_token_stream_with_cycler_instance(
-                prefix_token_stream.clone(),
+    if path.contains_variable() {
+        let variants = cycler.instances.iter().map(|instance| {
+            let instance_name = format_ident!("{}", instance.name);
+            let accessor_path = path_to_accessor_token_stream_with_cycler_instance(
+                prefix.clone(),
                 path,
                 reference_type,
-                Some(cycler_instance),
-            ));
-            token_stream_within_match.append(TokenTree::Punct(Punct::new(',', Spacing::Alone)));
+                Some(&instance.name),
+            );
+            quote! {
+                CyclerInstance::#instance_name => #accessor_path,
+            }
+        });
+        quote! {
+            match instance {
+                #(#variants)*
+            }
         }
-        token_stream.append(TokenTree::Group(Group::new(
-            Delimiter::Brace,
-            token_stream_within_match,
-        )));
-        token_stream
     } else {
-        path_to_accessor_token_stream_with_cycler_instance(
-            prefix_token_stream,
-            path,
-            reference_type,
-            None,
-        )
+        path_to_accessor_token_stream_with_cycler_instance(prefix, path, reference_type, None)
     }
 }
 
 fn path_to_accessor_token_stream_with_cycler_instance(
```

Comment:
This is a one of the success stories where a high code coverage leads to less effort in the review. :100:

Resolve if read.